pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,235 @@
+type LegacySATokenCleaner struct {
+	client           clientset.Interface
+	clock            clock.Clock
+	saLister         listersv1.ServiceAccountLister
+	saInformerSynced cache.InformerSynced
+
+	secretLister         listersv1.SecretLister
+	secretInformerSynced cache.InformerSynced
+
+	podLister         listersv1.PodLister
+	podInformerSynced cache.InformerSynced
+
+	syncInterval         time.Duration
+	minimumSinceLastUsed time.Duration
+}
+
+// NewLegacySATokenCleaner returns a new *NewLegacySATokenCleaner.
+func NewLegacySATokenCleaner(saInformer coreinformers.ServiceAccountInformer, secretInformer coreinformers.SecretInformer, podInformer coreinformers.PodInformer, client clientset.Interface, cl clock.Clock, options LegacySATokenCleanerOptions) (*LegacySATokenCleaner, error) {
+	tc := &LegacySATokenCleaner{
+		client:               client,
+		clock:                cl,
+		saLister:             saInformer.Lister(),
+		saInformerSynced:     saInformer.Informer().HasSynced,
+		secretLister:         secretInformer.Lister(),
+		secretInformerSynced: secretInformer.Informer().HasSynced,
+		podLister:            podInformer.Lister(),
+		podInformerSynced:    podInformer.Informer().HasSynced,
+		minimumSinceLastUsed: options.CleanUpPeriod,
+		syncInterval:         options.SyncInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.syncInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+
+	now := tc.clock.Now().UTC()
+	trackedSince, err := tc.getTrackedSinceDate(ctx)
+	if err != nil {
+		return
+	}
+
+	if now.Before(trackedSince.Add(tc.minimumSinceLastUsed)) {
+		// we haven't been tracking long enough
+		return
+	}
+
+	preserveCreatedOnOrAfter := now.Add(-tc.minimumSinceLastUsed)
+	preserveUsedOnOrAfter := now.Add(-tc.minimumSinceLastUsed).Format(dateFormat)
+
+	secretList, err := tc.secretLister.Secrets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		return
+	}
+
+	podMountedSecrets := make(map[string]sets.String)
+	for _, secret := range secretList {
+		if secret.Type != v1.SecretTypeServiceAccountToken {
+			continue
+		}
+		if !secret.CreationTimestamp.Time.Before(preserveCreatedOnOrAfter) {
+			continue
+		}
+
+		lastUsed, ok := secret.Labels[serviceaccount.LastUsedLabelKey]
+		if ok && lastUsed >= preserveUsedOnOrAfter {
```

Comment:
after the cheap check of existence and string comparison, should we check the lastUsed value is well-formed?

Reply:
I added one step to check whether the lastUsed can be parsed corrrectly.